optaplanner-core/src/main/java/org/optaplanner/core/impl/domain/solution/cloner/gizmo/GizmoSolutionClonerImplementor.java

```diff
@@ -0,0 +1,764 @@
+            });
+
+            gizmoMemberDescriptor.whenIsField(fd -> {
+                resultHandleHolder[0] = noCloneBranch.readInstanceField(fd, thisObj);
+            });
+
+            ResultHandle fieldValue = resultHandleHolder[0];
+            AssignableResultHandle cloneValue = noCloneBranch.createVariable(deeplyClonedField.getType());
+            writeDeepCloneInstructions(noCloneBranch, solutionInfo,
+                    deeplyClonedField.getType(), gizmoMemberDescriptor.getType(), fieldValue, cloneValue, createdCloneMap);
+
+            gizmoMemberDescriptor.whenIsField(fd -> {
+                noCloneBranch.writeInstanceField(fd, clone, cloneValue);
+            });
+            gizmoMemberDescriptor.whenIsMethod(md -> {
+                Optional<MethodDescriptor> maybeSetter = gizmoMemberDescriptor.getSetter();
+                if (!maybeSetter.isPresent()) {
+                    throw new IllegalStateException("Field (" + gizmoMemberDescriptor.getName() + ") of class (" +
+                            gizmoMemberDescriptor.getDeclaringClassName() + ") does not have a setter.");
+                }
+                gizmoMemberDescriptor.invokeMemberMethod(noCloneBranch, maybeSetter.get(), clone, cloneValue);
+            });
+        }
+        noCloneBranch.returnValue(clone);
+    }
+
+    /**
+     * Writes the following code:
+     *
+     * <pre>
+     * // If getter a field
+     * clone.member = original.member
+     * // If getter a method (i.e. Quarkus)
+     * clone.setMember(original.getMember());
+     * </pre>
+     * 
+     * @param methodCreator
+     * @param shallowlyClonedField
+     * @param thisObj
+     * @param clone
+     */
+    private static void writeShallowCloneInstructions(GizmoSolutionOrEntityDescriptor solutionInfo,
+            BytecodeCreator methodCreator, GizmoMemberDescriptor shallowlyClonedField,
+            ResultHandle thisObj, ResultHandle clone, ResultHandle createdCloneMap) {
+        try {
+            boolean isArray = shallowlyClonedField.getTypeName().endsWith("[]");
+            Class<?> type = null;
+            if (shallowlyClonedField.getType() instanceof Class) {
+                type = (Class<?>) shallowlyClonedField.getType();
+            }
+
+            List<Class<?>> entitySubclasses = Collections.emptyList();
+            if (type == null && !isArray) {
+                type = Class.forName(shallowlyClonedField.getTypeName().replace('/', '.'), false,
+                        Thread.currentThread().getContextClassLoader());
+            }
+
+            if (type != null && !isArray) {
+                entitySubclasses = solutionInfo.getSolutionDescriptor().getEntityClassSet().stream()
+                        .filter(type::isAssignableFrom).collect(Collectors.toList());
+            }
+
+            final List<Class<?>> finalEntitySubclasses = entitySubclasses;
+            final Class<?> finalType = type;
```

Comment:
I really don't like the fact that we're mixing final and non-final vars.
I know it's because of the lambdas, but there has to be a nicer way.
How about extracting this stuff to a separate method, where these would be the arguments?

Reply:
Done via adding read/write methods to GizmoMemberDescriptor